webapp/src/Controller/API/JudgehostController.php

```diff
@@ -1248,4 +1254,91 @@ private function getSubmissionsToJudge(Judgehost $judgehost, $restrictJudgingOnS
         $submissions = $queryBuilder->getQuery()->getResult();
         return $submissions;
     }
+
+    /**
+     * Get files for a given type and id.
+     * @Rest\Get("/get_file/{type}/{id}")
+     * @Security("is_granted('ROLE_JURY') or is_granted('ROLE_JUDGEHOST')")
+     * @param Request $request
+     * @param string  $type
+     * @param string  $id
+     * @return array
+     * @throws \Doctrine\ORM\NonUniqueResultException
+     * @SWG\Response(
+     *     response="200",
+     *     description="The files for the submission, testcase or script.",
+     *     @SWG\Schema(ref="#/definitions/SourceCodeList")
+     * )
+     * @SWG\Parameter(ref="#/parameters/id")
+     */
+    public function getFileAction(Request $request, string $type, string $id)
+    {
+        switch($type) {
+            case 'source':
+                return $this->getSourceFiles($id);
+            case 'testcase':
+                // TODO: decide whether testcase or input/output are the right granularity.
+                throw new BadRequestHttpException('Not yet implemented.');
+            case 'compile':
+            case 'run':
+            case 'compare':
+                return $this->getExecutableFiles($id);
+            default:
+                throw new BadRequestHttpException('Unknown type requested.');
+        }
+    }
+
+    private function getSourceFiles(string $id) {
+        $queryBuilder = $this->em->createQueryBuilder()
+            ->from(SubmissionFile::class, 'f')
+            ->join('f.submission', 's')
+            ->select('f, s')
```

Comment:
I think we can drop selecting s here? Maybe even not load the submission at all but use where f.submission = :submitid?
Same below for exec files

Reply:
Done.